ufbx.h

```diff
@@ -119,412 +205,2171 @@ struct ufbx_prop {
+// is `false` depending on how it's stored in the FBX file.
+typedef struct ufbx_vertex_attrib {
 	ufbx_real *data;
 	int32_t *indices;
-	size_t num_elements;
-};
+	size_t num_values;
+	size_t value_reals;
+	bool unique_per_vertex;
+} ufbx_vertex_attrib;
 
-struct ufbx_vertex_vec2 {
+// 1D vertex attribute, see `ufbx_vertex_attrib` for information
+typedef struct ufbx_vertex_real {
+	ufbx_real *data;
+	int32_t *indices;
+	size_t num_values;
+	size_t value_reals;
+	bool unique_per_vertex;
+} ufbx_vertex_real;
+
+// 2D vertex attribute, see `ufbx_vertex_attrib` for information
+typedef struct ufbx_vertex_vec2 {
 	ufbx_vec2 *data;
 	int32_t *indices;
-	size_t num_elements;
-};
+	size_t num_values;
+	size_t value_reals;
+	bool unique_per_vertex;
+} ufbx_vertex_vec2;
 
-struct ufbx_vertex_vec3 {
+// 3D vertex attribute, see `ufbx_vertex_attrib` for information
+typedef struct ufbx_vertex_vec3 {
 	ufbx_vec3 *data;
 	int32_t *indices;
-	size_t num_elements;
-};
+	size_t num_values;
+	size_t value_reals;
+	bool unique_per_vertex;
+} ufbx_vertex_vec3;
 
-struct ufbx_vertex_vec4 {
+// 4D vertex attribute, see `ufbx_vertex_attrib` for information
+typedef struct ufbx_vertex_vec4 {
 	ufbx_vec4 *data;
 	int32_t *indices;
-	size_t num_elements;
-};
+	size_t num_values;
+	size_t value_reals;
+	bool unique_per_vertex;
+} ufbx_vertex_vec4;
 
-struct ufbx_uv_set {
+// Vertex UV set/layer
+typedef struct ufbx_uv_set {
 	ufbx_string name;
 	int32_t index;
-	ufbx_vertex_vec2 vertex_uv;
-	ufbx_vertex_vec3 vertex_binormal;
-	ufbx_vertex_vec3 vertex_tangent;
-};
 
-struct ufbx_color_set {
+	// Vertex attributes, see `ufbx_mesh` attributes for more information
+	ufbx_vertex_vec2 vertex_uv;        // < UV / texture coordinates
+	ufbx_vertex_vec3 vertex_tangent;   // < (optional) Tangent vector in UV.x direction
+	ufbx_vertex_vec3 vertex_bitangent; // < (optional) Tangent vector in UV.y direction
+} ufbx_uv_set;
+
+// Vertex color set/layer
+typedef struct ufbx_color_set {
 	ufbx_string name;
 	int32_t index;
-	ufbx_vertex_vec4 vertex_color;
-};
 
-struct ufbx_edge {
+	// Vertex attributes, see `ufbx_mesh` attributes for more information
+	ufbx_vertex_vec4 vertex_color; // < Per-vertex RGBA color
+} ufbx_color_set;
+
+UFBX_LIST_TYPE(ufbx_uv_set_list, ufbx_uv_set);
+UFBX_LIST_TYPE(ufbx_color_set_list, ufbx_color_set);
+
+// Edge between two _indices_ in a mesh
+typedef struct ufbx_edge {
 	uint32_t indices[2];
+} ufbx_edge;
+
+// Polygonal face with arbitrary number vertices, a single face contains a 
+// contiguous range of mesh indices, eg. `{5,3}` would have indices 5, 6, 7
+//
+// NOTE: `num_indices` maybe less than 3 in which case the face is invalid!
+// [TODO #23: should probably remove the bad faces at load time]
+typedef struct ufbx_face {
+	uint32_t index_begin;
+	uint32_t num_indices;
+} ufbx_face;
+
+typedef struct ufbx_mesh_material {
+	ufbx_material *material;
+
+	size_t num_faces;
+	size_t num_triangles;
+	int32_t *faces;
```

Comment:
Just a question :)
What does `faces` point to? 
How do we iterate through all vertices in a part of a mesh associated with given material?

Reply:
I actually renamed that to `face_indices` recently in another (rather broken) branch and was not sure if it was a good idea, seems like it would be :D 

The `faces[0..num_faces]` is a list of indices to `ufbx_mesh.faces[]`, I didn't want to use `ufbx_face` structures directly as someone might want to access the `ufbx_mesh.face_smoothing[]` property that requires a face index.

Iterating through the _vertices_ that touch a material would be something like:

```c
ufbx_mesh *mesh;
ufbx_mesh_material *mesh_mat = &mesh->materials.data[mat_ix];
for (size_t face_ix = 0; face_ix < mesh_mat->num_faces; face_ix++) {
	ufbx_face face = mesh->faces[mesh_mat->faces[face_ix]];
	for (size_t i = 0; i = face.num_indices; i++) {
		size_t index = face.index_begin + i;
		// If you want just the index of the vertex
		int32_t vertex_ix = mesh->vertex_indices[index];
		// If you need the position as well
		ufbx_vec3 pos = mesh->vertices[vertex_ix];
		// .. or equivalently use the attribute getter
		ufbx_vec3 also_pos = ufbx_get_vertex_vec3(&mesh->vertex_position, index);
		// .. or get any other attribute
		ufbx_vec2 uv = ufbx_get_vertex_vec2(&mesh->vertex_uv, index);
	}
}
```

Reply:
I see, that makes sense. 
Yeah, I think a more descriptive name like 'face_indices would be better. Maybe also you could make it an unsigned int so that its usage is more clear?

Reply:
Yea I'll definitely keep the name `face_indices` then, I'm a bit unsure what to do about the signedness of all the index types. Right now they're pretty much all over the place. I personally prefer unsigned indices everywhere, but all the FBX indices (eg. `PolygonVertexIndex`, `Indices` in skin weights) are signed, so silently casting them to be unsigned feels wrong as well.

Now that I think about it maybe I should just force everything as unsigned as the FBX data is just raw bits so I can assume it to be either one (without breaking strict aliasing etc). Thanks for the feedback!